storage/innobase/dict/dict0dict.cc

```diff
@@ -1580,150 +1557,18 @@ dict_table_rename_in_cache(
-						= '\0';
-				}
-			}
-
-			strncpy(fkid, foreign->id, (sizeof fkid) - 1);
-			fkid[(sizeof fkid) - 1] = '\0';
-
-			const bool on_tmp = dict_table_t::is_temporary_name(
-				fkid);
-
-			if (!on_tmp) {
-				innobase_convert_to_filename_charset(
-					strchr(fkid, '/') + 1,
-					strchr(foreign->id, '/') + 1,
-					MAX_TABLE_NAME_LEN+20);
-			}
-
-			old_id = mem_strdup(foreign->id);
-
-			if (strlen(fkid) > strlen(old_name_cs_filename)
-			    + ((sizeof dict_ibfk) - 1)
-			    && !memcmp(fkid, old_name_cs_filename,
-				       strlen(old_name_cs_filename))
-			    && !memcmp(fkid + strlen(old_name_cs_filename),
-				       dict_ibfk, (sizeof dict_ibfk) - 1)) {
-
-				/* This is a generated >= 4.0.18 format id */
-
-				char	table_name[MAX_TABLE_NAME_LEN + 1];
-				uint	errors = 0;
-
-				if (strlen(table->name.m_name)
-				    > strlen(old_name)) {
-					foreign->id = static_cast<char*>(
-						mem_heap_alloc(
-						foreign->heap,
-						strlen(table->name.m_name)
-						+ strlen(old_id) + 1));
-				}
-
-				/* Convert the table name to UTF-8 */
-				strncpy(table_name, table->name.m_name,
-					MAX_TABLE_NAME_LEN);
-				table_name[MAX_TABLE_NAME_LEN] = '\0';
-				innobase_convert_to_system_charset(
-					strchr(table_name, '/') + 1,
-					strchr(table->name.m_name, '/') + 1,
-					MAX_TABLE_NAME_LEN, &errors);
-
-				if (errors) {
-					/* Table name could not be converted
-					from charset my_charset_filename to
-					UTF-8. This means that the table name
-					is already in UTF-8 (#mysql50#). */
-					strncpy(table_name, table->name.m_name,
-						MAX_TABLE_NAME_LEN);
-					table_name[MAX_TABLE_NAME_LEN] = '\0';
-				}
-
-				/* Replace the prefix 'databasename/tablename'
-				with the new names */
-				strcpy(foreign->id, table_name);
-				if (on_tmp) {
-					strcat(foreign->id,
-					       old_id + strlen(old_name));
-				} else {
-					sprintf(strchr(foreign->id, '/') + 1,
-						"%s%s",
-						strchr(table_name, '/') +1,
-						strstr(old_id, "_ibfk_") );
-				}
-
-			} else {
-				/* This is a >= 4.0.18 format id where the user
-				gave the id name */
-				db_len = dict_get_db_name_len(
-					table->name.m_name) + 1;
-
-				if (db_len - 1
-				    > dict_get_db_name_len(foreign->id)) {
-
-					foreign->id = static_cast<char*>(
-						mem_heap_alloc(
-						foreign->heap,
-						db_len + strlen(old_id) + 1));
-				}
-
-				/* Replace the database prefix in id with the
-				one from table->name */
-
-				memcpy(foreign->id,
-				       table->name.m_name, db_len);
-
-				strcpy(foreign->id + db_len,
-				       dict_remove_db_name(old_id));
-			}
-
-			ut_free(old_id);
+		size_t fklen = snprintf(nullptr, 0, "%s\377%s",
+					table->name.m_name,
+					foreign->sql_id());
+		char* id = foreign->id;
+		if (fklen > strlen(id)) {
+			id = static_cast<char*>(
+				mem_heap_alloc(foreign->heap, fklen + 1));
 		}
-
+		snprintf(id, fklen + 1, "%s\377%s",
+			 table->name.m_name, foreign->sql_id());
```

Comment:
Do we need to call foreign->sql_id() twice?

Reply:
Not really: we could cache the return value of the first call.

Reply:
d596884d4052c318dd92e0d3f3fdd6755f9b3519 is introducing another local variable `sql_id`. We must in fact avoid reassigning `foreign->id` before we have invoked `table->foreign_set.erase(it)`, because erasing would compare `foreign->id`, and the entry may need to be moved to a different position in the binary tree.